torch/csrc/jit/codegen/cuda/scheduler/utils.h

```diff
@@ -289,21 +290,85 @@ namespace matmul_utils {
 TORCH_CUDA_CU_API void scheduleContiguousVectorLoad(
     TensorView* tv,
     MatMulTileOptions tile,
-    int vector_word);
+    int vector_word,
+    bool vectorize = true);
 
 //! Schedule utility for mma output in matmul main loop:
 //!  Realize the hierarchical tiling based on the given tiling options.
+//! TODO: rewrite this one with makeTile
 TORCH_CUDA_CU_API void scheduleWarpTileWithReduction(
     TensorView* tv,
     MatMulTileOptions tile);
 
 //! Schedule utility for mma output in matmul main loop:
 //!  Realize the hierarchical tiling based on the given tiling options
 //! on consumers of mma ops in epilog.
+//! TODO: remove this one eventually.
 TORCH_CUDA_CU_API void scheduleWarpTileWithNoReduction(
     TensorView* tv,
     MatMulTileOptions tile);
 
+//! Lower level primitive spliting inner iterdomains into tiles:
+//! Eg.
+//!  A[B,I0,I1,I2] -> makeTile({1,2,3})
+//! Gives A[B, I0o, I1o, I2o, I0i(1), I1i(2), I2i(3)]
+TORCH_CUDA_CU_API void makeTile(TensorView* tv, std::vector<int> tile_sizes);
+
+//! Order the inner tile dimensions as the original order in
+//!  root domain. Also putting broadcast domains on the left.
+//! Eg. A[I0o,I1o,B2o,I0i,I1i,B2i] (root domain: I1,B,I0)
+//! -> A[I0o, I1o, B2o, B2i, I1i, I0i]
+//! This is used to facilitate data layout swizzling and
+//!  defining vectorized loads.
+TORCH_CUDA_CU_API void orderTiledConcreteIdAsRoot(TensorView* tv);
+
+//! Orders the root id ordering of the given tv as
+//! [Batch, Previous Reduction, M, N, K]
+//!  for easier processing of later scheduling steps.
+//!
+//! This matching works on root domain only, and
+//!  will throw if the tv has a leaf iterdomain that is
+//!  not a root id.
+TORCH_CUDA_CU_API void canonicalizeMmaTvOrdering(TensorView* tv);
+
+//! Propagate current transformations on from_tv to all tensorviews
+//!  within boundary.
+//! \param from_tv: target tv to replay transformations from.
+//! \param pos: position of target tv to replay to.
+//! \param upper_boundary:
+//!    boundary on the producer side where the propagation will stop.
+//!    an empty boundary means the propagator will **not** propagate
+//!    upward the DAG.
+//! \param lower_boundary: target tv to replay transformations from.
+//!    boundary on the consumer side where the propagation will stop.
+//!    an empty boundary means the propagator will **not** propagate
+//!    downward the DAG.
+//! \param propagate_parallel_type: Indicates if the propagator will
+//!    also replay parallel types on the matched iterdomains on the
+//!    replayed tensordomain.
+//! \param include_boundaries: Indicates if the tv's on the boundary
```

Comment:
```suggestion
//! \param include_boundaries: Indicates if the tvs on the boundary
```

Reply:
Please see updated logic in `BoundedDirectionalTransformPropagator`.